archive/android_img.ksy

```diff
@@ -0,0 +1,116 @@
+meta:
+  id: android_img
+  title: Android Boot Image
+  license: CC0-1.0
+  file-extension: img
+  endian: le
+doc-ref: https://source.android.com/devices/bootloader/boot-image-header
+seq:
+  - id: magic
+    contents: ANDROID!
+  - id: kernel
+    type: load
+  - id: ramdisk
+    type: load
+  - id: second
+    type: load
+  - id: tags_load
+    type: u4
+  - id: page_size
+    type: u4
+  - id: header_version
+    type: u4
+  - id: os_version
+    type: os_version
+  - id: name
+    type: strz
+    size: 16
+    encoding: ASCII
+  - id: cmdline
+    type: strz
+    size: 512
+    encoding: ASCII
+  - id: sha
+    size: 32
+  - id: extra_cmdline
+    type: strz
+    size: 1024
+    encoding: ASCII
+  - id: recovery_dtbo
+    type: load_offset
+    if: header_version > 0
+  - id: boot_header_size
+    type: u4
+    if: header_version > 0
+  - id: dtb
+    type: load_long
+    if: header_version > 1
+instances:
+  base:
+    value: kernel.addr - 0x00008000
+  kernel_offset:
+    value: kernel.addr - base
+  ramdisk_offset:
+    value: ramdisk.addr - base
+  second_offset:
+    value: second.addr - base
+  tags_offset:
+    value: tags_load - base
+  recovery_dtbo_offset:
+    value: recovery_dtbo.offset
+  dtb_offset:
+    value: dtb.addr - base
+    if: header_version > 1
+  kernel_img:
+    pos: page_size
+    size: kernel.size
+  ramdisk_img:
+    pos: ((page_size + kernel.size + page_size - 1) / page_size) * page_size
+    size: ramdisk.size
+    if: ramdisk.size > 0
+  second_img:
+    pos: ((page_size + kernel.size + ramdisk.size + page_size - 1) / page_size) * page_size
+    size: second.size
+    if: second.size > 0
+  recovery_dtbo_img:
+    pos: ((page_size + kernel.size + ramdisk.size + second.size + page_size - 1) / page_size) * page_size
+    size: recovery_dtbo.size
+    if: header_version > 0 and recovery_dtbo.size > 0
+  dtb_img:
+    pos: ((page_size + kernel.size + ramdisk.size + second.size + recovery_dtbo.size + page_size - 1) / page_size) * page_size
+    size: dtb.size
+    if: header_version > 1 and dtb.size > 0
+types:
+  load:
+    seq:
+      - id: size
+        type: u4
+      - id: addr
+        type: u4
+  load_long:
+    seq:
+      - id: size
+        type: u4
+      - id: addr
+        type: u8
+  load_offset:
+    seq:
+      - id: size
+        type: u4
+      - id: offset
+        type: u8
```

Comment:
I wonder how these types differ, whether they cannot be unified? I guess that `addr` is some absolute position, whereas `offset` is relative one, but I don't know. Is this semantic difference really so significant?

Reply:
> I guess that addr is some absolute position, whereas offset is relative one, but I don't know.

Exactly. This offset is the relative position for recovery_dtbo in boot.img, addr is a physical memory address to load the entity too

Here is how google extract recovery_dtbo from boot.img
https://android.googlesource.com/platform/system/core/+/refs/heads/android10-c2f2-release/mkbootimg/unpack_bootimg.py#113

Fixed.